File: 0040-combination-sum-ii/cpp/Solution.cpp
class Solution {
public:
    vector<vector<int>> result;
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        unordered_map<int, int> hashTable;
        for (int i = 0; i < candidates.size(); ++i) {
            auto it = hashTable.find(candidates[i]);
            if (it == hashTable.end()) {
                hashTable[candidates[i]] = 1;
            } else {
                hashTable[candidates[i]] = hashTable[candidates[i]] + 1;
            }
        }
        vector<int> nums;
        vector<int> counts;
        for (int i = 0; i < candidates.size(); ++i) {
            auto it = hashTable.find(candidates[i]);
            if (it != hashTable.end()) {
                nums.push_back(candidates[i]);
                counts.push_back(hashTable[candidates[i]]);
                hashTable.erase(it);
            }
        }
        vector<int> path;
        backtrack(nums, counts, 0, target, path);
        return result;
    }
    void backtrack(vector<int> nums, vector<int> counts, int k, int left, vector<int> &path) {
        if (left == 0) {
            result.push_back(path);
            return;
        }
        if (left < 0 || k == nums.size()) {
            return;
        }
        for (int count = 0; count <= counts[k]; ++count) {
            for (int i = 0; i < count; ++i) {
                path.push_back(nums[k]);
            }
            backtrack(nums, counts, k + 1, left - count * nums[k], path);
            for (int i = 0; i < count; ++i) {
                path.pop_back();
            }
        }
    }
};

File: 0040-combination-sum-ii/python/Solution.py
class Solution:
    def __init__(self):
        self.result = []

    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        hashTable = dict()
        for i in range(len(candidates)):
            if candidates[i] not in hashTable:
                hashTable[candidates[i]] = 1
            else:
                hashTable[candidates[i]] += 1
        nums = []
        counts = []
        for i in range(len(candidates)):
            if candidates[i] in hashTable:
                nums.append(candidates[i])
                counts.append(hashTable[candidates[i]])
                hashTable.pop(candidates[i])

        self.backtrack(nums, counts, 0, target, [])
        return self.result

    def backtrack(self, nums, counts, k, left, path):
        if left == 0:
            self.result.append(path[:])
            return
        if k == len(nums) or left < 0:
            return
        for count in range(counts[k] +1):
            for i in range(count):
                path.append(nums[k])
            self.backtrack(nums, counts, k+1, left - count*nums[k], path)
            for i in range(count):
                path.pop()